Source/JavaScriptCore/runtime/SetPrototype.cpp

```diff
@@ -558,6 +559,128 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncIsSubsetOf, (JSGlobalObject* globalObject,
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSSet* thisSet = getSet(globalObject, callFrame->thisValue());
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSValue otherValue = callFrame->argument(0);
+
+    if (otherValue.isCell()) [[likely]] {
+        if (auto* otherSet = jsDynamicCast<JSSet*>(otherValue.asCell())) [[likely]] {
+            if (setPrimordialWatchpointIsValid(vm, otherSet)) [[likely]] {
+                scope.release();
+                return fastSetIsSupersetOf(globalObject, thisSet, otherSet);
+            }
+        }
+    }
+
+    uint32_t otherSize = getSetSizeAsInt(globalObject, otherValue);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    ASSERT(otherValue.isObject());
+    JSObject* otherObject = asObject(otherValue);
+
+    JSValue has = otherObject->get(globalObject, vm.propertyNames->has);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!has.isCallable()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set.prototype.isSupersetOf expects other.has to be callable"_s);
+
+    JSValue keys = otherObject->get(globalObject, vm.propertyNames->keys);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!keys.isCallable()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set.prototype.isSupersetOf expects other.keys to be callable"_s);
+
+    if (thisSet->size() < otherSize)
+        return JSValue::encode(jsBoolean(false));
+
+    CallData keysCallData = JSC::getCallData(keys);
+    MarkedArgumentBuffer keysArgs;
+    ASSERT(!keysArgs.hasOverflowed());
+    JSValue keysResult = call(globalObject, keys, keysCallData, otherValue, keysArgs);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSValue nextMethod = keysResult.get(globalObject, vm.propertyNames->next);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!nextMethod.isCallable()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set.prototype.isSupersetOf expects other.keys().next to be callable"_s);
+
+    CallData nextCallData = JSC::getCallData(nextMethod);
+
+    std::optional<CachedCall> cachedNextCall;
+    if (nextCallData.type == CallData::Type::JS) [[likely]] {
+        cachedNextCall.emplace(globalObject, jsCast<JSFunction*>(nextMethod), 0);
+        RETURN_IF_EXCEPTION(scope, { });
+    }
+
+    while (true) {
+        JSValue nextResult;
+        if (cachedNextCall) [[likely]] {
+            nextResult = cachedNextCall->callWithArguments(globalObject, keysResult);
+            RETURN_IF_EXCEPTION(scope, { });
+        } else {
+            MarkedArgumentBuffer nextArgs;
+            ASSERT(!nextArgs.hasOverflowed());
+            nextResult = call(globalObject, nextMethod, nextCallData, keysResult, nextArgs);
+            RETURN_IF_EXCEPTION(scope, { });
+        }
+
+        JSValue done = nextResult.get(globalObject, vm.propertyNames->done);
+        RETURN_IF_EXCEPTION(scope, { });
+        if (done.toBoolean(globalObject))
+            break;
+        RETURN_IF_EXCEPTION(scope, { });
```

Comment:
This looks mot correct. If toBoolean does not throw, then we do not need it. If it is not, then this is required before this if statement (or using EXCEPTION_ASSERT)